--- gcc/config/ba/t-ba	2015-02-12 12:19:55.000000000 +0000
+++ gcc/config/ba/t-ba	2015-02-13 10:32:11.448961638 +0000
@@ -9,12 +9,12 @@
 
 # -----------------------------------------------------------------------------
 #special multilib for jn51xx (as it lacks some instructions)
-MULTILIB_OPTIONS  += mcpu=jn51xx
-MULTILIB_DIRNAMES += mcpu_jn51xx
+MULTILIB_OPTIONS  += mcpu=jn51xx|Os
+MULTILIB_DIRNAMES += mcpu_jn51xx_sizeopt
 
 #generic BA options
-MULTILIB_OPTIONS  += mle m16regs Os mno-hard-div|mno-hard-mul
-MULTILIB_DIRNAMES += le 16regs optsize min
+#MULTILIB_OPTIONS  += mle m16regs Os mno-hard-div|mno-hard-mul
+#MULTILIB_DIRNAMES += le 16regs optsize min
 
 # -----------------------------------------------------------------------------
 # Following exceptions bend build system into creating only two
